tokio-trace-fmt/src/filter/reload.rs

```diff
@@ -0,0 +1,149 @@
+use parking_lot::RwLock;
+use std::{
+    error, fmt,
+    sync::{Arc, Weak},
+};
+use tokio_trace_core::{subscriber::Interest, Metadata};
+use {filter::Filter, span::Context};
+
+#[derive(Debug)]
+pub struct ReloadFilter<F> {
+    inner: Arc<RwLock<F>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Handle<F> {
+    inner: Weak<RwLock<F>>,
+}
+
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+}
+
+#[derive(Debug)]
+enum ErrorKind {
+    SubscriberGone,
+}
+
+// ===== impl ReloadFilter =====
+
+impl<F, N> Filter<N> for ReloadFilter<F>
+where
+    F: Filter<N>,
+{
+    fn callsite_enabled(&self, _: &Metadata, _: &Context<N>) -> Interest {
+        // TODO(eliza): When tokio-rs/tokio#1039 lands, we can allow our
+        // interest to be cached. For now, we must always return `sometimes`.
+        Interest::sometimes()
+    }
+
+    fn enabled(&self, metadata: &Metadata, ctx: &Context<N>) -> bool {
+        self.inner.read().enabled(metadata, ctx)
+    }
+}
+
+impl<F: 'static> ReloadFilter<F> {
+    pub fn new<N>(f: F) -> Self
+    where
+        F: Filter<N>,
+    {
+        Self {
+            inner: Arc::new(RwLock::new(f)),
+        }
+    }
+
+    pub fn handle(&self) -> Handle<F> {
+        Handle {
+            inner: Arc::downgrade(&self.inner),
+        }
+    }
+}
+
+// ===== impl Handle =====
+
+impl<F: 'static> Handle<F> {
```

Comment:
would be nice to also have a `get_ref(&self) -> F` to get the current filter, so that something like get the current filters and add a new filter to it. (which is a use case I can think of).

Reply:
👍 we can do that!

Reply:
Hmm, actually, that's non-trivial since the filter is behind a `RwLock`. We could have a function taking a closure, or one that clones the current filter if it implements `Clone`...

Reply:
Could we get a closure that gets a `&mut Filter` or a `(Filter, Handle)`?

Reply:
added in 11158c663ad906d968cef935b467483efa869b56